### DISCUSSION TITLE
Results files in the repo

### DISCUSSION THREAD

**Discussion by tsmbland** [EnergySystemsModellingLab/MUSE_OS#351] [General]
I think it's worth thinking again about which files we're committing to the repo.

Currently we're committing all results files from all the example/tutorial models. This means that every time any of these models change, we get a whole load of changes (currently 25,000 additions and deletions in #349), and I don't think it's sustainable to keep committing all these changes to the repo. It's already 1.5 GB and growing.

As far as I can tell, the only reason that we need to commit the results files is for the regression tests. Would it not be better to create some kind of checksum? We'd have to be a bit careful because we often get subtly different results on different operating systems, so we'd need the checksum to be tolerant to some degree of error, but there must be some solution that doesn't involve committing the full results files...

@dalonsoa @alexdewar

**Comment by alexdewar** [comment]
I agree. It's not good to be changing 20k lines with every PR, for multiple reasons.

Thinking about it, we could change the regression tests so that you can pass in a path to some old results (with an env var or something), then we could drop the results from the repo. We can do some trickery with our CI workflows to first generate the old results from whatever version of MUSE is on the commit the PR is based on (e.g. the tip of `develop`) first. (We can even cache the results, if we're feeling fancy.) We could also upload generated results with the documentation, if that's needed.

**Reply by tsmbland** [reply]
Possibly, although what would we do when we legitimately want to change the results, as the tests would fail whenever there's a difference compared to the cached files?

**Reply by alexdewar** [reply]
Ok, I was feeling clever for a minute there :laughing:, but you're right that things would break in that case. D'oh.

I'm not saying this is a *good* option, but sometimes for tests which need large amounts of data, people put it in a separate repo, so at least the ugliness is somewhere else and you don't have to deal with it unless you're specifically running those tests. Sometimes people use git LFS with this (in our case we'd probably need to put the results into zip files to make it worth while): https://git-lfs.com/

**Reply by alexdewar** [reply]
Ok, new idea. We have a file which contains the commit hash of the last time the results were known to be good. If you deliberately change the results, then you also need to update this hash (we can provide a script to do this). Then we can use the approach I suggested above and compare results from the version pointed to with the hash with those from the current version.

**Reply by tsmbland** [reply]
I think that could work!

**Reply by alexdewar** [reply]
I can't find anywhere in the code where `example_outputs` is populated... Are we currently doing that manually?

**Reply by tsmbland** [reply]
Yeah (well that's what I've been doing at least)

**Reply by dalonsoa** [reply]
I agree that this is not sustainable, but I prefer the idea of keeping a file with checksums - platform dependent, if needed. Pointing to a specific hash feels a bit hacky and I think it might fail in forks or if history needs clean-up (as we might need to do to get rid of 1.5 Gb of useless data).

**Reply by alexdewar** [reply]
But if we did rebase the entire repo, we'd still only need to change that one commit hash rather than checksums for all of the individual files. I agree it's a bit hacky, but if we can relegate the generation/caching of regression data to the CI then it will make things easier.

Although the results would differ between platforms (e.g. mac ARM would be different), the bigger problem is that you'll get different results for different CPU types, so you would get different results on Intel and AMD processors, even though they're both x86. It also means if you're running this on x86 macOS or ARM Linux (ok, this one isn't too likely), then the tests will also fail. I think having a tolerance for floating point error is the right way to do this.

**Reply by alexdewar** [reply]
PS -- I just double-checked and the `.git` folder in my repo is ~40MB, which isn't terrible. Maybe you were also including your `.venv` in the count @tsmbland?

Either way we should fix this, but we probably don't need to retroactively rip all the results files out of the repo.

**Reply by tsmbland** [reply]
Ahh you're right - silly me!

**Comment by tsmbland** [comment]
I'm going to close this. The outputs are now much smaller than they once were, so I'm less worried about the repository booming. I also find it quite useful committing the results files as it's easy to see exactly what's changed in the diffs, which you wouldn't get with a checksum
